doc/introduction/returns.rst

```diff
@@ -0,0 +1,182 @@
+
+    Your issue may be because:
+
+    * You are calculating the Jacobian of the QNode using the :ref:`NumPy <numpy_interf>` or
+      :ref:`TensorFlow <tf_interf>` interface. For example, the following will now raise an error:
+
+      .. code-block:: python
+
+          from pennylane import numpy as np
+
+          dev = qml.device("default.qubit", wires=1)
+
+          @qml.qnode(dev)
+          def circuit(x):
+              qml.RX(x, wires=0)
+              return qml.expval(qml.PauliY(0)), qml.expval(qml.PauliZ(0))
+
+          x = np.array(0.5, requires_grad=True)
+          qml.jacobian(circuit)(x)
+
+      Follow the instructions :ref:`here <return-autograd-tf-gotcha>` to fix this issue.
+      Alternatively, consider porting your code to use the :ref:`JAX <jax_interf>` or
+      :ref:`Torch <torch_interf>` interface, which could unlock additional features and performance
+      benefits!
+
+    * You are returning differently-shaped quantities together, such as
+      :func:`expval() <pennylane.expval>` and :func:`probs() <pennylane.probs>`. For example, the
+      following code is compatible with version 0.29 of PennyLane but will raise an error in version
+      0.30 and above:
+
+      .. code-block:: python
+
+          dev = qml.device("default.qubit", wires=1)
+
+          @qml.qnode(dev)
+          def circuit(x):
+              qml.RX(x, wires=0)
+              return qml.expval(qml.PauliZ(0)), qml.probs(0)
+
+          def result(x):
+              expval, p0, p1 = circuit(x)
+              return expval + p0 - p1
+
+          x = np.array(0.5, requires_grad=True)
+          result(x)
+
+      Such issues can be addressed by updating how the return of a QNode is processed, being aware
+      of unpacking, slicing, and indexing. The example above would be fixed simply by updating
+      ``result()`` to:
+
+      .. code-block:: python
+
+          def result(x):
+              expval, (p0, p1) = circuit(x)
+              return expval + p0 - p1
+
+.. details::
+    :title: I am a device developer
+    :href: device-developer
+
+    If you are a :ref:`device developer <plugin_overview>`, your issue may be because:
+
+    * Your device inherits from :class:`~.QubitDevice` and you have overriden or interact with the
+      :meth:`~.QubitDevice.execute`, :meth:`~.QubitDevice.batch_execute`, or
+      :meth:`~.QubitDevice.statistics` methods. Your device will need to be updated to accommodate
+      for the new return behaviour. An example can be found
+      `here <https://github.com/PennyLaneAI/pennylane-qiskit/pull/281>`_ for the
+      `Qiskit plugin <https://docs.pennylane.ai/projects/qiskit/en/latest/>`_.
+
+    * Your device inherits from :class:`Device <pennylane.Device>`, in which case you may need to rewrite the
+      :meth:`~.QubitDevice.execute`, :meth:`~.QubitDevice.batch_execute`, and
+      :meth:`~.QubitDevice.statistics` methods. Please
+      `reach out to us <https://discuss.pennylane.ai>`_ for guidance!

```

Comment:
@rmoyard, anything else we should add here?

Reply:
It is good! I think here it be worth to mention that it only concerns non qubit devices.

Reply:
I updated it to "Your device inherits directly from Device", does that work?

Reply:
Yes great!